cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -401,6 +418,45 @@ canonDbPath NodeConfiguration{ncDatabaseFile = DbFile dbFp} = do
   createDirectoryIfMissing True fp
   return fp
 
+#ifdef UNIX
+-- | Make sure the VRF private key file is readable only
+-- by the current process owner the node is running under.
+checkVRFFilePermissions :: Maybe FilePath -> ExceptT VRFPrivateKeyFilePermissionError IO ()
+checkVRFFilePermissions mFp =
+  case mFp of
+    Just vrfPrivKey -> do
+      fs <- liftIO $ getFileStatus vrfPrivKey
+      let fm = fileMode fs
+
+      -- Check the the VRF private key file does not give read/write/exec permissions to others.
+      when (hasOtherPermissions fm)
+           (left $ OtherPermissionsExist vrfPrivKey)
+      -- Check the the VRF private key file does not give read/write/exec permissions to any group.
+      when (hasGroupPermissions fm)
+           (left $ GroupPermissionsExist vrfPrivKey)
+      -- Check that the VRF private key file has owner read permissions.
+      unless (hasOwnerPermissions fm)
+             (left $ OwnerDoesNotHaveReadWrite vrfPrivKey)
+
+    Nothing -> right ()
+ where
+
+  hasPermission :: FileMode -> FileMode -> Bool
+  hasPermission fModeA fModeB = fModeA `intersectFileModes` fModeB == fModeB
+
+  hasOwnerPermissions :: FileMode -> Bool
+  hasOwnerPermissions fm' = fm' `hasPermission` ownerReadMode
+
+  hasOtherPermissions :: FileMode -> Bool
+  hasOtherPermissions fm' =
+    any (fm' `hasPermission`) [otherReadMode, otherWriteMode, otherExecuteMode]
+
+  hasGroupPermissions :: FileMode -> Bool
+  hasGroupPermissions fm' =
+    any (fm' `hasPermission`) [groupReadMode, groupWriteMode, groupExecuteMode]
```

Comment:
In principle these tests can be simplified by using bit operations. One can union the three modes together and then test all of them in one go. The test for all uses intersection and tests if the result is non-0 (not == fModeB).